### ISSUE TITLE
Support m1 chipset

### ISSUE THREAD

**Issue by mooreds** [FusionAuth/fusionauth-issues#1523]
### Problem
M1 chipset is not supported.

### Solution

Support it.

### Alternatives/workarounds

Rosetta, perhaps? https://support.apple.com/en-us/HT211861

### Additional context

Pulled over from: https://github.com/FusionAuth/fusionauth-containers/issues/73

### Community guidelines
All issues filed in this repository must abide by the [FusionAuth community guidelines](https://fusionauth.io/community/forum/topic/1000/code-of-conduct).

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by robotdan** [comment]
This should work as of 1.32.0 in theory. 
https://fusionauth.io/docs/v1/tech/release-notes/#version-1-32-0

When we start up on macOS, we download Java 17 for `x86_64` or `arm64` architecture. As far as I know with this change, Rosetta should not be required. 

Prior to version `1.32.0` we were only installing Java for `x86_64` on macOS. So in this case, I think it will still work, but it would run through Rosetta. 

We should be getting our M1 MacBooks shortly so we can confirm this behavior. If anyone else has an M1 MacBook and wants to try running `1.32.0` let us know how it goes! 

Starting in 2022 we will be developing on M1 so, I can guarantee it will work! We would also like to begin to support AWS Graviton and other ARM based servers. 

According to this article ( https://dev.to/aws-builders/aws-graviton-processors-3nk3 ) the Linux arch type on Graviton is `aarch64` - so it may already work. But we have not tested it.

**Comment by mooreds** [comment]
cc @dmitryzan

**Comment by dmitryzan** [comment]
@mooreds Unfortunately the official 1.32.0 nor 1.32.1 versions don't work for me. The same symptoms: high CPU load (100-300% for the java process inside the container according to top) and freezes. But the `jerryhopper/fusionauth-app:1.32.0` works perfect.

Think it's not about the java version you pull inside the container but the container itself. It should be built for aarch64 with multi-arch.

If you attach to the running container and run uname the official image shows
```
$ uname -a
Linux 83eb24b25f44 5.10.76-linuxkit #1 SMP PREEMPT Mon Nov 8 11:22:26 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
when the Jerry's multi-arch version:
```
$ uname -a
Linux 0edb7bb3349c 5.10.76-linuxkit #1 SMP PREEMPT Mon Nov 8 11:22:26 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by robotdan** [comment]
Thanks for the details and assistance @dmitryzan - this is helpful. 

So you are running the Docker image on macOS and because we are building the image on `linux/x86_64`, it sounds like Docker is likely running this all through Rosetta which is crushing your CPU. 

So then the fix for M1 and Docker in general is to not only use the correct Java version (in the Docker image) but to publish an additional Docker image built on `linux/arm64` in addition to our current published image.

**Comment by dmitryzan** [comment]
Thank you Daniel!
I'm not a big expert in building images for multiple architectures but seems you can configure this as a single build process with Multi-Arch builds https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/
(also think the community images and the Jerry's build already use it, as another example).

**Comment by robotdan** [comment]
Hacking around a bit, I think I have it working. Feel free to test it out. I published tag `1.32.1_multi` with `linux/amd64` and `linux/arm64` images. 

https://hub.docker.com/layers/fusionauth/fusionauth-app/1.32.1_multi/images/sha256-03f95a54f11561c44a2995a42034319b37faeef873f9390433e9884013ccdfb3?context=repo

Jerry is also publishing a 3rd image for `linux/arm/v7`, I tried that and it didn't work. I think I have a few kinks to work out to publish all of the versions Jerry is doing, but I'm getting close.

**Comment by dmitryzan** [comment]
`1.32.1_multi` works much better! Great work Daniel, thank you!
Ran a few tests and clicked through a number of admin pages, no 100% CPU load anymore (20% at most occasionally, think it was always this way).
Will run more next week.

**Comment by robotdan** [comment]
Great, I think we are close. Once we iron out a few more things, we should be able to start publishing these additional images with each release.

**Comment by ceefour** [comment]
Linking to starter thread https://github.com/FusionAuth/fusionauth-containers/issues/49 :)

**Comment by jerryhopper** [comment]
> @mooreds Unfortunately the official 1.32.0 nor 1.32.1 versions don't work for me. The same symptoms: high CPU load (100-300% for the java process inside the container according to top) and freezes. But the `jerryhopper/fusionauth-app:1.32.0` works perfect. Running on MBP 2021 with M1 Max 32GB, MacOS Monterey 12.1, Docker for Mac v4.3.1 (72247).
> 
> Think it's not about the java version you pull inside the container but the container itself. think should be built for `linux/arm64`.
> 
> If you attach to the running container and run uname the official image shows
> 
> ```
> $ uname -a
> Linux 83eb24b25f44 5.10.76-linuxkit #1 SMP PREEMPT Mon Nov 8 11:22:26 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
> ```
> 
> when the Jerry's multi-arch version:
> 
> ```
> $ uname -a
> Linux 0edb7bb3349c 5.10.76-linuxkit #1 SMP PREEMPT Mon Nov 8 11:22:26 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
> ```
> 
> (but the container itself is `linux/arm64`)

This is correct, it seems the image was compiled using the wrong kernel-headers. (compiled on the wrong arch)
As im using github buildX action, it automatically compiles all archs with the correct kernelheaders  - i dont see that in the savant-build script, and lack knowledge of savant.  Also i tried to find info about multiarch building using savant, but couldnt find any usable info.

But, ive worked on a github workflow file that works with the original FA dockerfile (with minor tweaks) and informed robotdan about it.   Essentially it does the same as savantbuild - only better :)

Maybe there is a specific reason to use savant, and not use github-ci, idk.

**Comment by robotdan** [comment]
> Maybe there is a specific reason to use savant, and not use github-ci, idk.


Not really, all we are doing in our Savant build scripts is calling `docker build`, it is just a glorified `ant` script. It sounds like using GH actions is likely the right choice here.

**Comment by rscheuermann** [comment]
Thanks so much @robotdan ! I'm going to test the latest `1.32.1_multi` image for about a week. I was regularly experiencing runaway memory leaks and core dumps on the Jerry's image. Your image already seems faster and more stable.

**Comment by rscheuermann** [comment]
@robotdan Is there anything blocking `1.32.1_multi` from being pushed as the standard `latest` build? Do you have an ETA? 

So far, after 2 weeks of running `1.32.1_multi`, we've not encountered any crashes.

**Comment by robotdan** [comment]
This test image should have support for 3 arch types. If this works, next release will include all images. 
`fusionauth/fusionauth-app:1.32.1_multi_test_1`

**Comment by robotdan** [comment]
I didn't want to leave out the IBM i, p and z series servers. So I added a few more arch types. Not sure if we'll publish these going forward or not, but we could. 😁

![image](https://user-images.githubusercontent.com/1148709/151721404-75afdf55-62cf-43eb-9fcd-9c83af95e839.png)

**Comment by robotdan** [comment]
https://hub.docker.com/r/fusionauth/fusionauth-app/tags

Released `arm64` (and others) back to `1.24.0` and all upcoming releases will include multi-arch. 

And the fast path install will already support macOS.
